File: iOS_demo/UyaliBeautyFaceSDKDemo-OC/UyaliBeautyFaceSDKDemo-OC/美颜/FaceBeautyView.m
//
//  FaceBeautyView.m
//  UyaliBeautyFaceSDKDemo-OC
//
//  Created by S weet on 2023/2/27.
//

#import "FaceBeautyView.h"
#import "UyaliButton.h"

@interface FaceBeautyView ()

@property (nonatomic, strong)NSArray *beautyItems;

@property (nonatomic, assign) NSInteger currentTag;

@end

@implementation FaceBeautyView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self initDatas];
        [self setupView];
    }
    return self;
}

- (void)initDatas {
    _beautyItems = @[
        @{@"image": @"beauty_skin_white", @"name": @"美白"},
        @{@"image": @"beauty_skin_abrade", @"name": @"磨皮"},
        @{@"image": @"beauty_bright_eye", @"name": @"亮眼"},
        @{@"image": @"beauty_bright_teeth", @"name": @"白牙"}];
    
    _currentTag = 100;
}

#pragma mark - Setter & Getter
- (void)setCurrentTag:(NSInteger)currentTag {
    if (_currentTag != currentTag) {
        UyaliButton *newBtn = (UyaliButton *)[self viewWithTag:currentTag];
        UyaliButton *oldBtn = (UyaliButton *)[self viewWithTag:_currentTag];
        
        newBtn.selected = YES;
        oldBtn.selected = NO;
        
        if (currentTag == 100) {//美白 done 0 - 100
            _slider.minimumValue = 0;
            _slider.maximumValue = 100;
        } else if (currentTag == 101) {//磨皮 done 0 - 100
            _slider.minimumValue = 0;
            _slider.maximumValue = 100;
        } else if (currentTag == 102) {//亮眼 done 0 - 100
            _slider.minimumValue = 0;
            _slider.maximumValue = 100;
        } else if (currentTag == 103) {//白牙 done 0 - 100
            _slider.minimumValue = 0;
            _slider.maximumValue = 100;
        }
        
        _currentTag = currentTag;
    }
}

#pragma mark - UI
- (void)setupView {
    self.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent:0.4];
    
    _slider = [[UISlider alloc]initWithFrame:CGRectMake(16, 24, self.bounds.size.width-32, 30)];
    [self addSubview:_slider];
    [_slider addTarget:self action:@selector(sliderValueChanged:) forControlEvents:UIControlEventValueChanged];
    
    _valueLabel = [[UILabel alloc]initWithFrame:CGRectMake(0, 0, self.bounds.size.width, _slider.frame.origin.y)];
    [self addSubview:_valueLabel];
    _valueLabel.font = [UIFont systemFontOfSize:15 weight:UIFontWeightMedium];
    _valueLabel.textColor = [UIColor whiteColor];
    _valueLabel.textAlignment = NSTextAlignmentCenter;
    _valueLabel.text = @"0.0";
    
    UIScrollView *scroll = [[UIScrollView alloc]initWithFrame:CGRectMake(0, 60, self.bounds.size.width, self.bounds.size.height-70)];
    [self addSubview:scroll];
    
    CGFloat buttonWidth = scroll.bounds.size.height;
    CGFloat buttonHeight = scroll.bounds.size.height;
    CGFloat buttonMargin = 16;
    
    scroll.contentSize = CGSizeMake(buttonMargin+(buttonWidth+buttonMargin)*_beautyItems.count, scroll.frame.size.height);
    
    for (NSInteger i = 0; i < _beautyItems.count; i++) {
        NSDictionary *item = _beautyItems[i];
        UyaliButton *button = [[UyaliButton alloc]initWithFrame:CGRectMake(buttonMargin+(buttonMargin+buttonWidth)*i, 0, buttonWidth, buttonHeight)];
        [scroll addSubview:button];
        [button setTitle:item[@"name"] forState:UIControlStateNormal];
        [button setImage:[UIImage imageNamed:item[@"image"]] forState:UIControlStateNormal];
        [button addTarget:self action:@selector(beautyButtonAction:) forControlEvents:UIControlEventTouchUpInside];
        button.tag = 100+i;
        if (i == 0) {
            button.selected = YES;
            
            _slider.minimumValue = 0;
            _slider.maximumValue = 100;
        }
        button.layer.cornerRadius = buttonWidth/2;
    }
}

#pragma mark - Action
- (void)beautyButtonAction:(UyaliButton *)button {
    self.currentTag = button.tag;
    if (self.delegate != nil && [self.delegate respondsToSelector:@selector(changeBeautyType:)]) {
        [self.delegate changeBeautyType:button.tag];
    }
}

- (void)sliderValueChanged:(UISlider *)currentSlider {
    _valueLabel.text = [NSString stringWithFormat:@"%.1f",currentSlider.value];
    if (self.delegate != nil && [self.delegate respondsToSelector:@selector(getBeautyDeltaVale:type:)]) {
        [self.delegate getBeautyDeltaVale:_slider.value type:_currentTag];
    }
}

@end

File: iOS_demo/UyaliBeautyFaceSDKDemo/UyaliBeautyFaceSDKDemo/FaceBeautyView.swift
//
//  FaceBeautyView.swift
//  UyaliBeautyFaceSDKDemo
//
//  Created by S weet on 2023/2/16.
//

import UIKit

@objc protocol FaceBeautyDelegate: AnyObject {
    func changeBeautyType(type: Int)
    func getBeautyDeltaValue(value:Float, type:Int)
}

class FaceBeautyView: UIView {
    
    weak var delegate: FaceBeautyDelegate!

    var slider: UISlider!
    var valueLabel: UILabel!
    
    private let beautyItems: [[String: String]] = [
        ["image": "beauty_skin_white", "name": "美白"],
        ["image": "beauty_skin_abrade", "name": "磨皮"]
    ]
    
    private var currentTag: Int = 100 {
        didSet {
            if currentTag != oldValue {
                let newBtn = viewWithTag(currentTag) as! UyaliButton
                let oldBtn = viewWithTag(oldValue) as! UyaliButton
                
                newBtn.isSelected = true
                oldBtn.isSelected = false
                
                if currentTag == 100 {//美白 done 0 - 100
                    slider.minimumValue = 0
                    slider.maximumValue = 100
                } else if currentTag == 101 {//磨皮 done 0 - 100
                    slider.minimumValue = 0
                    slider.maximumValue = 100
                }
            }
        }
    }

    override init(frame: CGRect) {
        super.init(frame: frame)
        setupView()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    //MARK: UI
    func setupView() {
        self.backgroundColor = .black.withAlphaComponent(0.4)
        
        slider = UISlider(frame: CGRect(x: 16, y: 24, width: bounds.width-32, height: 30))
        addSubview(slider)
        slider.addTarget(self, action: #selector(sliderValueChanged(currentSlider: )), for: .valueChanged)
        
        valueLabel = UILabel(frame: CGRect(x: 0, y: 0, width: bounds.width, height: slider.frame.origin.y))
        addSubview(valueLabel)
        valueLabel.font = .systemFont(ofSize: 15, weight: .medium)
        valueLabel.textColor = .white
        valueLabel.textAlignment = .center
        valueLabel.text = "0.0"
        
        let scroll = UIScrollView(frame: CGRect(x: 0, y: 60, width: bounds.width, height: bounds.height-70))
        addSubview(scroll)
        
        let buttonWidth = scroll.bounds.height
        let buttonHeight = scroll.bounds.height
        let buttonMargin = 16.0
        
        scroll.contentSize = CGSize(width: buttonMargin+(buttonWidth+buttonMargin)*CGFloat(beautyItems.count), height: scroll.frame.height)
        
        for i in 0..<beautyItems.count {
            let item = beautyItems[i]
            let button = UyaliButton(frame: CGRect(x: buttonMargin+(buttonMargin+buttonWidth)*CGFloat(i), y: 0, width: buttonWidth, height: buttonHeight))
            scroll.addSubview(button)
            button.setTitle(item["name"], for: .normal)
            button.setImage(UIImage(named: item["image"]!), for: .normal)
            button.addTarget(self, action: #selector(beautyButtonAction(button:)), for: .touchUpInside)
            button.tag = 100+i
            if i == 0 {
                button.isSelected = true
                currentTag = 100
                
                slider.minimumValue = 0
                slider.maximumValue = 100
            }
            button.layer.cornerRadius = buttonWidth/2
        }
    }
    
    //MARK: Action
    @objc func beautyButtonAction(button: UyaliButton) {
        currentTag = button.tag
        if delegate != nil {
            delegate.changeBeautyType(type: button.tag)
        }
    }
    
    @objc func sliderValueChanged(currentSlider: UISlider) {
        valueLabel.text = String(format: "%.1f", currentSlider.value)
        if delegate != nil {
            delegate.getBeautyDeltaValue(value: slider.value, type: currentTag)
        }
    }
}
